### DISCUSSION TITLE
Next.js 13 build generates HTML files with a .html.html file extension

### DISCUSSION THREAD

**Discussion by fanmau** [vercel/next.js#47337] [Help] [answered]
### Summary

# Next.js 13 build generates HTML files with a .html.html file extension

My website is generating HTML static files with an additional .html file extension in the build process.

The homepage file name is normal.

```
https://fanmau.github.io/index.html
```
The article pages generated file names with an additional `.html` file extension.
```
https://fanmau.github.io/blog/2023/1.html.html
https://fanmau.github.io/blog/2023/2.html.html
```

## Demo

[https://fanmau.github.io](https://fanmau.github.io)

## Source code

Source code repository： [GitHub Pages](https://github.com/fanmau/fanmau.github.io) 

Website HTML static files： [https://github.com/fanmau/fanmau.github.io/tree/fanma](https://github.com/fanmau/fanmau.github.io/tree/fanma) 

## Site structure

This is the website structure I want.
```
https://fanmau.github.io/index.html
https://fanmau.github.io/blog/2023/1.html
https://fanmau.github.io/blog/2023/2.html
https://fanmau.github.io/blog/2023/3.html
https://fanmau.github.io/blog/2023/4.html
https://fanmau.github.io/blog/2023/5.html
```

But the generated file structure looks like this.

```
https://fanmau.github.io/index.html
https://fanmau.github.io/blog/2023/1.html.html
https://fanmau.github.io/blog/2023/2.html.html
https://fanmau.github.io/blog/2023/3.html.html
https://fanmau.github.io/blog/2023/4.html.html
https://fanmau.github.io/blog/2023/5.html.html
```
![FireShot Capture 003 - fanmau github io_blog_2023 at fanma · fanmau_fanmau github io - github com](https://user-images.githubusercontent.com/127578292/226510414-615cec5d-d000-4dde-8f4f-9d860ec8a7fa.png)

Actually, `https://fanmau.github.io/blog/2023/1.html` is also accessible normally.

## Dependency version information

```
  "dependencies": {
    "@types/node": "18.14.4",
    "@types/react": "18.0.28",
    "@types/react-dom": "18.0.11",
    "@types/slug": "^5.0.3",
    "date-fns": "^2.29.3",
    "gray-matter": "^4.0.3",
    "highlight.js": "^11.7.0",
    "moment-timezone": "^0.5.41",
    "next": "13.2.3",
    "next-sitemap": "^4.0.5",
    "prismjs": "^1.29.0",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "remark": "^14.0.2",
    "remark-html": "^15.0.1",
    "remark-prism": "^1.3.6",
    "sitemap": "^7.1.1",
    "slug": "^8.2.2",
    "typescript": "4.9.5"
  },
  "devDependencies": {
    "@types/prismjs": "^1.26.0",
    "@types/remark-prism": "^1.3.4"
  }
```

## I want to do this.

I want to keep the `.html` suffix in the params, but during file building, generate only `1.html` instead of `1.html.html` as it is now.

```
export const getStaticPaths: GetStaticPaths = async () => {
  return {
    paths: [
      { params: { slug: ["path", "to", "page", "5.html"] } },
      { params: { slug: ["another", "path", "1.html"] } },
      { params: { slug: ["another", "path", "8.html"] } },
      { params: { slug: ["another", "path", "9.html"] } },
    ],
    fallback: false,
  };
};
```

How can I remove the extra `.html` file extension that appears during file building?

### Example

https://github.com/fanmau/fanmau.github.io

**Comment by icyJoseph** [comment]
Hi,

The problem comes from doing `concat('.html')`, this function for example: `getAllPostIds`, returns a list of these:

```js
 return filenames.map(filename => ({
        params: {
          slug: [subdirectory, filename.replace(/\.md$/, '').concat('.html')]
        }
      }))
```

And there the slug is resolving to: `slug: ["bar", "foo.html"]`

Later when the builder encounters, for example `[...slug].tsx`, it creates a path: `/blog/bar/foo.html` (this is the path), and then when creating the output file, it does `/blog/bar/foo.html.html`, similarly for the root, you have somewhere else `.concat("html")`, and then the builder appends the `.html` to the path.

So, just remove the `.concat("html")` calls from your [lib/blog](https://github.com/fanmau/fanmau.github.io/blob/main/lib/blog.ts) helpers - I know it is odd that Next.js doesn't pick up on the presence of `.html` already, but it is what it is...

**Reply by fanmau** [reply]
Thank you for your answer.

I know that I added a concat('.html') here, and I just wanted this kind of effect.

```ts
 return filenames.map(filename => ({
        params: {
          slug: [subdirectory, filename.replace(/\.md$/, '').concat('.html')]
        }
      }))
```

I don't want to remove `concat('.html')` here, and I want to keep it as it is now. I spent three days trying many ways, but I couldn't find a better solution.

Today, I finally found a relatively good solution. I tested it several times today and found that it works well.

We can use the `fs.readdir()`, `path.join()`, and `fs.rename()` methods provided by Node.js's file system module "fs" and path module to rename all files with the ".html.html" suffix to files with the ".html" suffix. Please refer to the following sample code for the specific implementation method:

```js
const fs = require('fs');
const path = require('path');

function renameFiles(dir) {
  fs.readdir(dir, (err, files) => {               
    if (err) throw err;
    files.forEach(file => {                       
      const filePath = path.join(dir, file);      
      fs.stat(filePath, (err, stat) => {          
        if (err) throw err;
        if (stat.isFile() && file.endsWith('.html.html')) { 
          const newFilePath = path.join(dir, file.replace('.html.html', '.html'));
          fs.rename(filePath, newFilePath, (err) => {   
            if (err) throw err;
            console.log(`${filePath} renamed to ${newFilePath}`);
          });
        } else if (stat.isDirectory()) {              
          renameFiles(filePath);
        }
      });
    });
  });
}

renameFiles('path/to/directory');  
```

An example of the effect of using this method is shown below:

```ssh
out\blog\2023\1.html.html renamed to out\blog\2023\1.html
out\blog\2023\2.html.html renamed to out\blog\2023\2.html
out\blog\2023\3.html.html renamed to out\blog\2023\3.html
out\blog\2023\4.html.html renamed to out\blog\2023\4.html
out\blog\2023\5.html.html renamed to out\blog\2023\5.html
out\blog\2023\6.html.html renamed to out\blog\2023\6.html
out\blog\2023\7.html.html renamed to out\blog\2023\7.html
out\blog\2023\8.html.html renamed to out\blog\2023\8.html
out\blog\2023\9.html.html renamed to out\blog\2023\9.html
```

This is my current solution and it still works very well.

Now, all files with the duplicate `.html.html` extension have been renamed to `.html`.

**Reply by icyJoseph** [reply]
Mmm, you probably have your reasons to concat `.html` to the params -> slug property, but wouldn't it be more helpful to change your data structure:

```js
filenames.map((filename) => ({
  params: {
    slug: [subdirectory, filename.replace(/\.md$/, "")],
    filename: [subdirectory, filename.replace(/\.md$/, "").concat(".html")],
  },
}));
```

And then use the slug, or filename as needed through out your application? You can also shorten `filename.replace(/\.md$/, "").concat(".html")` to `filename.replace(/\.md$/, ".html")` 


Anyway, glad that you found some way to make it work for your app. Though I don't think it is the right way necessarily.

**Comment by fanmau** [comment, marked as answer]
Today, I finally found a relatively good solution. I tested it several times today and found that it works well.

We can use the `fs.readdir()`, `path.join()`, and `fs.rename()` methods provided by Node.js's file system module "fs" and path module to rename all files with the ".html.html" suffix to files with the ".html" suffix. Please refer to the following sample code for the specific implementation method:

```js
const fs = require('fs');
const path = require('path');

function renameFiles(dir) {
  fs.readdir(dir, (err, files) => {               
    if (err) throw err;
    files.forEach(file => {                       
      const filePath = path.join(dir, file);      
      fs.stat(filePath, (err, stat) => {          
        if (err) throw err;
        if (stat.isFile() && file.endsWith('.html.html')) { 
          const newFilePath = path.join(dir, file.replace('.html.html', '.html'));
          fs.rename(filePath, newFilePath, (err) => {   
            if (err) throw err;
            console.log(`${filePath} renamed to ${newFilePath}`);
          });
        } else if (stat.isDirectory()) {              
          renameFiles(filePath);
        }
      });
    });
  });
}

renameFiles('path/to/directory');  
```

An example of the effect of using this method is shown below:

```ssh
out\blog\2023\1.html.html renamed to out\blog\2023\1.html
out\blog\2023\2.html.html renamed to out\blog\2023\2.html
out\blog\2023\3.html.html renamed to out\blog\2023\3.html
out\blog\2023\4.html.html renamed to out\blog\2023\4.html
out\blog\2023\5.html.html renamed to out\blog\2023\5.html
out\blog\2023\6.html.html renamed to out\blog\2023\6.html
out\blog\2023\7.html.html renamed to out\blog\2023\7.html
out\blog\2023\8.html.html renamed to out\blog\2023\8.html
out\blog\2023\9.html.html renamed to out\blog\2023\9.html
```

This is my current solution and it still works very well.

Now, all files with the duplicate `.html.html` extension have been renamed to `.html`.
